rspace/src/main/scala/coop/rchain/rspace/InMemoryStore.scala

```diff
@@ -53,10 +51,28 @@ class InMemoryStore[T, C, P, A, K](
   private[this] val refine        = Map("path" -> "inmem")
   private[this] val entriesGauge  = Kamon.gauge(MetricsSource + ".entries").refine(refine)
 
-  private[rspace] def updateGauges() =
-    withTxn(createTxnRead())(_.readState { state =>
-      entriesGauge.set(state.size)
-    })
+  private[rspace] def createTxnReadF(): F[Transaction] = syncF.delay(createTxnRead())
+
+  private[rspace] def createTxnWriteF(): F[Transaction] = syncF.delay(createTxnWrite())
+
+  private[rspace] def withTxnFlatF[R](txnF: F[Transaction])(f: Transaction => F[R]): F[R] =
+    for {
+      txn    <- txnF
+      retErr <- f(txn).attempt
+      _ <- syncF.delay(retErr.map { _ =>
+            txn.readState({ state =>
+              entriesGauge.set(state.size)
+            })
+            txn.commit()
```

Comment:
Can the transaction be committed before updating the gauge in order to not potentially block the single writer thread?